File: 2024/09/part1.py
import fileinput


def read_input():
    for line in fileinput.input():
        for idx, ch in enumerate(line.strip()):
            file_id = -1 if idx & 1 else idx // 2
            for _ in range(int(ch)):
                yield file_id


def solve(disk):
    i = 0
    j = len(disk) - 1
    while i < j:
        if disk[i] == -1:
            if disk[j] != -1:
                disk[i] = disk[j]
                disk[j] = -1
                i += 1
            j -= 1
        else:
            i += 1
    print(sum(idx * file_id for idx, file_id in enumerate(disk) if file_id != -1))


if __name__ == "__main__":
    solve(list(read_input()))

File: 2024/09/part1.rs
use std::{env, fs::File, io::{self, BufRead}, path::Path};

fn read_input(path: &Path) -> io::Result<Vec<Option<usize>>> {
    let file = File::open(path)?;

    let mut disk = Vec::new();
    for line in io::BufReader::new(file).lines() {
        for (idx, ch) in line?.chars().enumerate() {
            let num_blocks = ch.to_digit(10).expect("digit");
            let file_id = if (idx & 1) == 0 { Some(idx / 2) } else { None };
            for _ in 0..num_blocks {
                disk.push(file_id);
            }
        }
    }

    Ok(disk)
}


fn solve(mut disk: Vec<Option<usize>>) {
    let mut i = 0;
    let mut j = disk.len() - 1;
    while i < j {
        match disk[i] {
            Some(_) => i += 1,
            None => {
                if disk[j].is_some() {
                    disk[i] = disk[j];
                    disk[j] = None;
                    i += 1;
                }
                j -= 1;
            }
        }
    }

    let checksum: usize = disk.iter().enumerate().filter_map(|(block_id, content)|
        match content {
            Some(file_id) => Some(block_id * *file_id),
            None => None,
        }
    ).sum();
    println!("{checksum}");
}

fn main() {
    for arg in env::args().skip(1) {
        println!("=== {arg}");
        let disk = read_input(Path::new(&arg)).expect("reading input");
        solve(disk);
    }
}
